links/links_test.go

```diff
@@ -0,0 +1,290 @@
+				"",
+				"9090",
+				"/prefix",
+				"http://localhost:8080/",
+			},
+		}
+
+		for _, tt := range tests {
+			du, err := url.Parse(tt.defaultURL)
+			So(err, ShouldBeNil)
+
+			h := http.Header{}
+			if tt.fwdProto != "" {
+				h.Add("X-Forwarded-Proto", tt.fwdProto)
+			}
+			if tt.fwdHost != "" {
+				h.Add("X-Forwarded-Host", tt.fwdHost)
+			}
+			if tt.fwdPort != "" {
+				h.Add("X-Forwarded-Port", tt.fwdPort)
+			}
+			if tt.fwdPathPrefix != "" {
+				h.Add("X-Forwarded-Path-Prefix", tt.fwdPathPrefix)
+			}
+
+			du.JoinPath()
+			builder := FromHeadersOrDefault(&h, du)
+			So(builder, ShouldNotBeNil)
+			So(builder.URL, ShouldNotBeNil)
+			So(builder.URL.String(), ShouldEqual, tt.want)
+
+		}
+
+	})
+
+}
+
+func TestBuilder_BuildLink(t *testing.T) {
+
+	Convey("Given a list of test cases", t, func() {
+		tests := []struct {
+			builderURL string
+			oldLink    string
+			want       string
+		}{
+			// Empty old link
+			{
+				"http://localhost:8080/",
+				"",
+				"http://localhost:8080/",
+			},
+			// Old link with no path
+			{
+				"http://localhost:8080/",
+				"http://localhost:8080/",
+				"http://localhost:8080/",
+			},
+			// Old link with different base url
+			{
+				"http://localhost:8080/",
+				"https://oldHost:1000/",
+				"http://localhost:8080/",
+			},
+			// Old link with path
+			{
+				"http://localhost:8080/",
+				"http://localhost:8080/some/path",
+				"http://localhost:8080/some/path",
+			},
+			// Old link with path and different base url
+			{
+				"http://localhost:8080/",
+				"http://oldHost:1000/some/path",
+				"http://localhost:8080/some/path",
+			},
+			// Old link without base url
+			{
+				"http://localhost:8080/",
+				"/some/path",
+				"http://localhost:8080/some/path",
+			},
+			// Old link with query params
+			{
+				"http://localhost:8080/",
+				"http://localhost:8080/some/path?param1=value1&param2=value2",
+				"http://localhost:8080/some/path?param1=value1&param2=value2",
+			},
+			// Old external link to new internal url
+			{
+				"http://localhost:8080/",
+				"https://some.api.host/v1/some/path",
+				"http://localhost:8080/some/path",
+			},
+			// Old external link to new external url
+			{
+				"https://some.api.host/v1",
+				"https://some.api.host/v1/some/path",
+				"https://some.api.host/v1/some/path",
+			},
+			// Old internal link to new external url
+			{
+				"https://some.api.host/v1",
+				"http://localhost:8080/some/path",
+				"https://some.api.host/v1/some/path",
+			},
+			// Old internal link to new external url with query params
+			{
+				"https://some.api.host/v1",
+				"http://localhost:8080/some/path?param1=value1&param2=value2",
+				"https://some.api.host/v1/some/path?param1=value1&param2=value2",
+			},
+		}
+
+		for _, tt := range tests {
+
+			bu, err := url.Parse(tt.builderURL)
+			So(err, ShouldBeNil)
+			builder := &Builder{URL: bu}
+
+			newurl, err := builder.BuildLink(tt.oldLink)
+			So(err, ShouldBeNil)
+			So(newurl, ShouldEqual, tt.want)
+
```

Comment:
There is a test here to check that there are no errors returned, but could we add some tests to cover any errors that might be thrown if an incorrect URL was provided for example?

Reply:
changes made